lib/db/observed.go

```diff
@@ -114,9 +114,18 @@ func (db *Lowlevel) RemovePendingFolder(id string, device []byte) {
 	}
 }
 
-// PendingFolders drops any invalid entries from the database as a side-effect.
-func (db *Lowlevel) PendingFolders() (map[string]map[protocol.DeviceID]ObservedFolder, error) {
-	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingFolder})
+// PendingFolders drops any invalid entries from the database as a side-effect.  The
+// returned results are filtered by the given device ID if it is non-empty.
+func (db *Lowlevel) PendingFolders(device []byte) (map[string]map[protocol.DeviceID]ObservedFolder, error) {
```

Comment:
Lets use `protocol.DeviceID` same as in all the other functions here. I'd generally always use a concrete type over `[]byte` if possible, eliminates a source of mistakes.

Reply:
That is an array type of fixed length, thus cannot be empty.  The `protocol.EmptyDeviceID` alternative is more of a semantic thing in my view, and I tried to keep the DB layer functions more focused on mechanics than semantics.

If you insist, I'll change it.  I'm just often uncertain whether your comments are a definite "no go, please change" or rather "is there a good reason for this?"  Because usually it's a conscious decision based on *my view* of "good" reasons ;-)

Reply:
My comments are usually meant like: "I think it should be done like this, because `insert argument`". The implicit request is to change it or provide arguments why it shouldn't be changed.

As I said, the entire db layer that you introduced (observed.go) uses `protocol.DeviceID`. Having it as such, makes it impossible to e.g. do `PendingFolders(folder)` where `folder` is a `[]byte`. I don't understand what your point is with fixed size arrays: `DeviceID` is a fixed size array underneath, yes, the argument right now is a slice.

Reply:
The argument was about the empty value. Slice can be nil (easy), while they array needs a special sentinel value, `protocol.EmptyDeviceID`.

Reply:
It's not a sentinel value, it's a convenience variable:`protocol.EmptyDeviceID == protocol.DeviceID{}`. Anyways that's irrelevant. `protocol.DeviceID` please.